src/ray/object_manager/object_manager.cc

```diff
@@ -306,17 +307,119 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
 }
 
 ray::Status ObjectManager::Cancel(const ObjectID &object_id) {
-  ray::Status status = object_directory_->UnsubscribeObjectLocations(object_id);
+  ray::Status status = object_directory_->UnsubscribeObjectLocations(
+      object_directory_pull_callback_id_, object_id);
   return status;
 }
 
-ray::Status ObjectManager::Wait(const std::vector<ObjectID> &object_ids,
-                                uint64_t timeout_ms, int num_ready_objects,
+ray::Status ObjectManager::Wait(const std::vector<ObjectID> &object_ids, int64_t wait_ms,
+                                uint64_t num_required_objects, bool wait_local,
                                 const WaitCallback &callback) {
-  // TODO: Implement wait.
+  UniqueID wait_id = UniqueID::from_random();
+
+  if (wait_ms < 0) {
+    return ray::Status::Invalid("Unable to wait negative wait time.");
+  }
+
+  if (wait_local) {
+    return ray::Status::NotImplemented("Wait for local objects is not yet implemented.");
+  }
+
+  RAY_CHECK(object_ids.size() != 0);
+  RAY_CHECK(num_required_objects != 0);
+  RAY_CHECK(num_required_objects <= object_ids.size());
+
+  // Initialize fields.
+  active_wait_requests_.emplace(wait_id, WaitState(*main_service_, wait_ms, callback));
+  auto &wait_state = active_wait_requests_.find(wait_id)->second;
+  wait_state.num_required_objects = num_required_objects;
+  wait_state.start_time = boost::posix_time::second_clock::local_time();
+  for (auto &oid : object_ids) {
+    if (local_objects_.count(oid) > 0) {
+      wait_state.found.insert(oid);
+    } else {
+      wait_state.remaining.insert(oid);
+    }
+  }
+
+  if (wait_state.found.size() >= wait_state.num_required_objects) {
+    // Requirements already satisfied.
+    WaitComplete(wait_id);
+  } else {
+    if (wait_ms == 0) {
+      for (auto &oid : wait_state.remaining) {
+        // Subscribe to object notifications.
+        wait_state.requested_objects.insert(oid);
+        object_directory_->LookupLocations(
+            oid, [this, wait_id](const std::vector<ClientID> &client_ids,
+                                 const ObjectID &object_id) {
+              auto &wait_state = active_wait_requests_.find(wait_id)->second;
+              if (!client_ids.empty()) {
+                wait_state.remaining.erase(object_id);
+                wait_state.found.insert(object_id);
+              }
+              wait_state.requested_objects.erase(object_id);
+              if (wait_state.requested_objects.empty()) {
```

Comment:
Why not return on the same check of `(wait_state.found.size() >= wait_state.num_required_objects)`? If it's because you're waiting for all of the results to return from the object table, that's probably not necessary. If we can, let's try to keep the logic the same between the `Lookup` and the `Subscribe` codepaths. It'd be best if we could use a common helper function between the two.

Also, if we change that check, we could probably get rid of the `requested_objects` member.

Reply:
We ought to keep `requested_objects` as this makes the code easier to read and reason about. Let's talk offline about the other changes.

Reply:
Okay, I'm fine with keeping `requested_objects`, but please add a comment for why you're completing the `Wait` here, since it's not obvious to the reader why the condition is different from the `Subscribe` codepath.